### PR TITLE
Project Happy Thoughts API - Anna Lindgren

### PR DESCRIPTION
**Pull request by magicvilla**
Hi!

I just want to inform that i have a delete endpoint for deleting a thought which i do not use at the moment. I want to implement this during the week thats why i have kept it there :) 

https://no-bad-thoughts-inc.netlify.app/
https://anna-happy-thoughts.herokuapp.com/

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -1,29 +1,119 @@
...
+app.use(express.json())
 
-// Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send(listEndpoints(app))
+})
+
+app.get('/thoughts', async (req, res) => {
+  const allThoughts = await Thought.find().sort({ createdAt: -1 }).limit(20).exec()
```
Comment (Frivridna) [inline server.js:46]: Nice that you added the .exec() method here! As it gives you a better stack tracing (and changes the thenable into a promise).

File: server.js [hunk 2]
```diff
@@ -1,29 +1,119 @@
...
 import cors from 'cors'
 import mongoose from 'mongoose'
+import listEndpoints from 'express-list-endpoints'
+import dotenv from 'dotenv'
+
+dotenv.config()
 
 const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/happyThoughts"
-mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
+// eslint-disable-next-line max-len
```
Comment (Frivridna) [inline server.js:11]: Why did you add this here? :)

File: server.js [hunk 3]
```diff
@@ -1,29 +1,119 @@
...
-//
-//   PORT=9000 npm start
-const port = process.env.PORT || 8080
+const thougthSchema = new mongoose.Schema({
+  message: {
+    type: String,
+    required: [true, 'Empty message not allowed'],
+    trim: true,
+    minlength: [5, 'Message must be longer than 5 characters'],
+    maxlength: [140, 'Message must be shorter than 140 characters']
```
Comment (Frivridna) [inline server.js:21]: Great that you are sending these messages between line 18-21 as additional information, as well as trimming whitespaces!

File: server.js [hunk 4]
```diff
@@ -1,29 +1,119 @@
...
+  hearts: {
+    type: Number,
+    default: 0
+  },
+  createdAt: {
+    type: Date,
+    default: Date.now
+  }
+})
+
```
Comment (Frivridna) [inline server.js:32]: Here you would add the author/username :)

**Review by Frivridna — APPROVED** [review]
Great backend Anna, and your frontend looks fab as well! 💯I am looking forward to the implementation of author and delete on back- and frontend when you have the time to do it.🙂 

One thing that I noticed is that you are only using the .exec() method once in your code. If you think it is a good practice and want to keep it, you can consider doing it in every query so that it is consistent throughout the code. It is just a suggestion🙂 But maybe you had a specific reason to only add it in the endpoint that sorts and limits? 

Your code is clean and easy to read. Looking forward to seeing your future projects! 🌟

File: README.md [hunk 5]
```diff
@@ -1,13 +1,25 @@
...
-Start by briefly describing the assignment in a sentence or two. Keep it short and to the point.
+## The assignment
+I started out by creating a thoughtSchema to define the structure of the document. I added default values and validators for the posted thought. 
 
-## The problem
+ Path | Description |
+| --- | --- |
+|GET "/thoughts" | Endpoint to get all thoughts. I used .sort() and .limit() to get the newest thoughts and limit to 20 thoughts. |
+| POST  "/thoughts" | Endpoint to post thoughts which expects JSON body with a thought message. I added validation to require a message, minimum message length of 5 characters and maximum length of 140 characters. If message does not meet the validation requirements, appropriate error message will be given.|
+| POST "/thoughts/:id/likes" | Endpoint to increase heart likes on a specific thought id. |
```
Comment (KarolCislowski) [inline README.md:16]: Nice documentation 🤓

File: server.js (hunk 3 above)
Comment (KarolCislowski) [inline server.js:21]: Nicely configured field 🤓

File: server.js [hunk 6]
```diff
@@ -1,29 +1,119 @@
...
+      { 
+        $inc: 
+          { 
+            hearts: 1 
+          } 
+      }, 
+      { 
+        new: true 
+      }
+    )
```
Comment (KarolCislowski) [inline server.js:89]: 💯

File: server.js [hunk 7]
```diff
@@ -1,29 +1,119 @@
...
+    const deletedThought = await Thought.findOneAndDelete({ _id: id })
+    if (deletedThought) {
+      res.json(deletedThought)
+    } else {
+      res.status(404).json({ message: 'Not found' })
+    }
+  } catch (error) {
+    res.status(400).json({ message: 'Invalid request', error })
+  }
 })
```
Comment (KarolCislowski) [inline server.js:115]: Remember to restrict access to this endpoint because not all users should have access to this functionality 🤓

**Review by KarolCislowski — APPROVED** [review]
Great API 🦄
